File: src/main/java/site/neworld/objective/mixin/MixinChunkStorage.java
package site.neworld.objective.mixin;

import com.mojang.datafixers.DataFixer;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.NbtIo;
import net.minecraft.util.math.ChunkPos;
import net.minecraft.world.FeatureUpdater;
import net.minecraft.world.storage.StorageIoWorker;
import net.minecraft.world.storage.VersionedChunkStorage;
import org.spongepowered.asm.mixin.Final;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Overwrite;
import org.spongepowered.asm.mixin.Shadow;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
import site.neworld.objective.IMixinChunkStorageTweakProviderConsumer;
import site.neworld.objective.MixinChunkStorageTweakProvider;

import java.io.ByteArrayOutputStream;
import java.io.DataOutputStream;
import java.io.File;
import java.io.IOException;

@Mixin(VersionedChunkStorage.class)
public class MixinChunkStorage {
    @Shadow
    @Final
    private StorageIoWorker worker;
    @Shadow
    private FeatureUpdater featureUpdater;

    public final MixinChunkStorageTweakProvider theStorage = new MixinChunkStorageTweakProvider();

    @Inject(method = "<init>", at = @At("RETURN"))
    public void init(File file, DataFixer dataFixer, boolean bl, CallbackInfo ci) throws IOException {
        worker.close(); // not null, but closed setting to null will lead to crashes
        theStorage.setup(file);
        if (this instanceof IMixinChunkStorageTweakProviderConsumer) {
            ((IMixinChunkStorageTweakProviderConsumer)this).consumeIMixinChunkStorageTweakProvider(theStorage);
        }
    }

    @Overwrite
    public CompoundTag getNbt(ChunkPos pos) {
        return theStorage.getNbtNow(new site.neworld.objective.utils.ChunkPos(pos.x, pos.z));
    }

    @Overwrite
    public void setTagAt(ChunkPos pos, CompoundTag compoundTag) {
        theStorage.setNbtNow(new site.neworld.objective.utils.ChunkPos(pos.x, pos.z), compoundTag);
        if (this.featureUpdater != null) {
            this.featureUpdater.markResolved(pos.toLong());
        }
    }

    @Overwrite
    public void completeAll() {
    }

    @Overwrite
    public void close() {
        theStorage.destroy();
    }
}

File: src/main/java/site/neworld/objective/mixin/MixinChunkStorage.kt
package site.neworld.objective.mixin

import com.mojang.datafixers.DataFixer
import kotlinx.coroutines.runBlocking
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.NbtIo
import net.minecraft.util.math.ChunkPos
import net.minecraft.world.FeatureUpdater
import net.minecraft.world.storage.StorageIoWorker
import net.minecraft.world.storage.VersionedChunkStorage
import org.spongepowered.asm.mixin.*
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo
import site.neworld.objective.Controller
import site.neworld.objective.data.LevelStorage
import java.io.*

@Mixin(VersionedChunkStorage::class)
open class MixinChunkStorage {
    @Shadow
    @Final
    private var worker: StorageIoWorker? = null

    @Shadow
    private val featureUpdater: FeatureUpdater? = null
    private var storage: LevelStorage? = null

    @Suppress("UNUSED_PARAMETER")
    @Inject(method = ["<init>"], at = [At("RETURN")])
    fun init(file: File, dataFixer: DataFixer?, bl: Boolean, ci: CallbackInfo) {
        // TODO(this is absolutely not a good way to do things, fix it)
        worker?.close()
        worker = null
        storage = Controller.open(file)
    }

    @Overwrite
    @Throws(IOException::class)
    fun getNbt(pos: ChunkPos) = runBlocking {
        val bytes = storage!!.get(site.neworld.objective.utils.ChunkPos(pos.x, pos.z))
        if (bytes != null) NbtIo.read(DataInputStream(ByteArrayInputStream(bytes))) else null
    }

    @Overwrite
    fun setTagAt(pos: ChunkPos, compoundTag: CompoundTag) {
        runBlocking {
            val outputStream = ByteArrayOutputStream()
            DataOutputStream(outputStream).use { NbtIo.write(compoundTag, it) }
            storage!!.set(site.neworld.objective.utils.ChunkPos(pos.x, pos.z), outputStream.toByteArray())
        }
        featureUpdater?.markResolved(pos.toLong())
    }

    @Overwrite
    fun completeAll() {} // TODO(currently there is no way to do this with objective engine)

    @Overwrite
    @Throws(IOException::class)
    open fun close() = storage!!.close()
}